rust/cyclotron-core/src/worker.rs

```diff
@@ -106,47 +129,74 @@ impl Worker {
+    /// may use to await the flushing of the updated job state, which happens asynchronously
+    /// to allow for batching of updates. Callers may drop the flush handle without impacting
+    /// the flushing of the update. This function returns an error if the caller tries to release
+    /// a job that this `Worker` doesn't know about, or if the worker tries to release a job
+    /// without having provided a next state for it.
+    ///
+    /// The flush handle returned here will resolve to an error if the asynchronous flush operation
+    /// fails in non-retryable fashion. Retryable errors during flush are not surfaced to the handle,
+    /// and the flush will be retried until it succeeds, a non-retryable error is encountered (e.g.
+    /// this workers lock on the job has been lost), or until the deadline is exceeded, if one is
+    /// provided. All updates will have at least one flush attempt.
+    pub fn release_job(&self, job_id: Uuid, deadline: Option<Duration>) -> FlushHandle {
         let update = {
-            let mut pending = self.pending.lock().unwrap();
-            let update = pending
-                .remove(&job_id)
-                .ok_or(QueueError::UnknownJobId(job_id))?;
-            // It's a programming error to flush a job without setting a new state
+            let mut running = self.running.lock().unwrap();
+            let Some(update) = running.remove(&job_id) else {
+                return FlushHandle::immediate(Err(JobError::UnknownJobId(job_id)));
+            };
             match update.state {
                 Some(JobState::Running) | None => {
-                    // Keep track of any /other/ updates that might have been stored, even in this case,
-                    // so a user can queue up the appropriate state transition and flush properly
-                    pending.insert(job_id, update);
-                    return Err(QueueError::FlushWithoutNextState(job_id));
+                    // Keep track of any /other/ updates that might have been stored, so this
+                    // error is recoverable simply by providing an appropriate new state.
+                    running.insert(job_id, update);
+                    return FlushHandle::immediate(Err(JobError::FlushWithoutNextState(job_id)));
                 }
                 _ => update,
             }
         };
-        let mut connection = self.pool.acquire().await?;
-        flush_job(&mut *connection, job_id, update).await
+
+        // If we were given a deadline, this update should be flushed at least as soon as then,
+        // otherwise we can wait the full linger time before flushing it.
+        let now = Utc::now();
+        let flush_by = now + deadline.unwrap_or(self.linger);
+        let deadline = deadline.map(|d| now + d);
+
+        let (pending, handle) = PendingUpdate::new(job_id, update, deadline);
+
+        let mut batch = self.flush_batch.lock().unwrap();
+        batch.add(pending, flush_by);
```

Comment:
Do we check if the batch is fully released at this point? Thinking that it could be all jobs are released but it continues to wait for the next flush timeout?